// @rakered/email/lib/templates/Blocks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Footer = exports.Header = exports.Code = exports.Title = exports.Content = exports.Container = exports.CallToAction = exports.Paragraph = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const Email_1 = require("./Email");
const Layout_1 = require("./Layout");
function Paragraph({ children }) {
    const style = Email_1.useTheme('paragraph');
    return jsx_runtime_1.jsx("div", Object.assign({ style: style }, { children: children }), void 0);
}
exports.Paragraph = Paragraph;
function CallToAction({ href, children }) {
    const style = Email_1.useTheme('callToAction');
    return (jsx_runtime_1.jsx("div", Object.assign({ style: style.outer }, { children: jsx_runtime_1.jsx("a", Object.assign({ href: href, style: style.inner }, { children: children }), void 0) }), void 0));
}
exports.CallToAction = CallToAction;
function Container({ children }) {
    const style = Email_1.useTheme('container');
    return (jsx_runtime_1.jsx("div", Object.assign({ style: style.outer }, { children: jsx_runtime_1.jsx("div", Object.assign({ style: style.inner }, { children: jsx_runtime_1.jsx(Layout_1.Grid, { children: children }, void 0) }), void 0) }), void 0));
}
exports.Container = Container;
function Content({ children }) {
    const style = Email_1.useTheme('content');
    return (jsx_runtime_1.jsx("div", Object.assign({ style: style.outer }, { children: jsx_runtime_1.jsx("div", Object.assign({ style: style.inner }, { children: jsx_runtime_1.jsx(Layout_1.Grid, { children: children }, void 0) }), void 0) }), void 0));
}
exports.Content = Content;
function Title({ children }) {
    const style = Email_1.useTheme('title');
    return jsx_runtime_1.jsx("div", Object.assign({ style: style }, { children: children }), void 0);
}
exports.Title = Title;
function Code({ children }) {
    const style = Email_1.useTheme('code');
    return (jsx_runtime_1.jsx("div", Object.assign({ style: style.outer }, { children: jsx_runtime_1.jsx("div", Object.assign({ style: style.inner }, { children: children }), void 0) }), void 0));
}
exports.Code = Code;
function Header({ logo, action, }) {
    const style = Email_1.useTheme('header');
    return (jsx_runtime_1.jsx("div", Object.assign({ style: style.outer }, { children: jsx_runtime_1.jsxs("div", Object.assign({ style: style.inner }, { children: [logo ? jsx_runtime_1.jsx("img", { style: style.logo, src: logo }, void 0) : null,
                action ? (jsx_runtime_1.jsx("a", Object.assign({ href: action.url, style: style.button }, { children: action.label }), void 0)) : null] }), void 0) }), void 0));
}
exports.Header = Header;
function Footer({ children }) {
    const style = Email_1.useTheme('footer');
    return (jsx_runtime_1.jsx("div", Object.assign({ style: style.outer }, { children: jsx_runtime_1.jsx("div", Object.assign({ style: style.inner }, { children: children }), void 0) }), void 0));
}
exports.Footer = Footer;
